api/client/service.proto

```diff
@@ -37,10 +39,57 @@ message ExecutionResponse {
   }
 }
 
+message RequestStreamRequest {
+  uint64 amount = 1;
+}
+
+message Condition {
+  enum ConditionType {
+    EQUALS = 0;
+    LESS_THEN = 1;
+    LESS_OR_EQUAL = 2;
+    GREATER_THEN = 3;
+    GREATER_OR_EQUAL = 4;
+  }
+
+  string name = 1;
+  ConditionType type = 2;
+}
+
+message Expectation {
+  enum DataType {
+    TEXT = 0;
+    INT = 1;
+  }
+  string name = 1;
+  string input_header = 2;
+  DataType input_header_data_type = 3;
+  Condition condition = 4;
+}
+
+message RequestStreamResponse {
+  // method/path/uri/authority/content_length can all be expressed in request headers.
+  // so technically they are redundant. In case of ambiguity headers may end up being duplicated.
+  // Note that as of today, the target upstream/cluster that will be used to replay traffic to
+  // fixes the destination address/port
+  google.protobuf.StringValue method = 1;
+  google.protobuf.StringValue uri = 2; // XXX(oschaaf): path?
+  google.protobuf.StringValue authority = 3;
+  envoy.api.v2.core.HeaderMap headers = 4;
+  google.protobuf.UInt32Value content_length = 5;
+
+  repeated Expectation expectations = 6;
+
+  // TODO(oschaaf): nice to have
+  // google.protobuf.StringValue sni_hostname = 10;
+}
```

Comment:
How far is this from the original replay format we discussed when we started this exercise? if you can share an abstract (i.e. non Google specific) idea of this, it would be good to compare.

Reply:
The draft replay format we came up with looked like:
 
``` proto
message KeyValue {
  string key = 1;
  string value = 2;
}

string request_method = 1;
string request_url = 2;
string authority = 3;
repeated KeyValue request_headers = 4;
string request_body = 5; (for POSTs)
 
optional int expected_response_code;
optional int expected_response_size; // (possibly also some to bound response size being at least N bytes or at most M bytes)
optional string expected_response_body;
optional KeyValue expected_response_headers; // (I assume subset of response headers that must be present)
 
// Possibly:
string sni_hostname = 6;
```